### ISSUE TITLE
The performance of using multiple gloo Context in multiple threads

### ISSUE THREAD

**Issue by flymark2010** [facebookincubator/gloo#42]
Hi, thanks for your great work.   
I've found the multiple Context use in multiple threads here:https://github.com/facebookincubator/gloo/issues/35 and tried this. But I found that we can't benefit much from this.  My code starts 4 threads to handle `allreduce sum` operation on 20 x 4MB data, compare with 4x20x4MB data in 1 thread,  on an 8 cores 32 threads E5 machine, with `GeForce GTX TITAN X`  .  And I use `mpirun`  starts 4 processes.  The result:  
```
is_gpu=true, time for [4] thread: 625.00 ms, time for single thread: 827.00 ms.
is_gpu=false, time for [4] thread: 312.00 ms, time for single thread: 362.00 ms.
```
Does I use multiple gloo Context in a wrong way? Or there are some limits on multiple threads using?   

Here is my code:  
```
#include <cassert>
#include <iostream>
#include <memory>
#include <gflags/gflags.h>
#include <glog/logging.h>

#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <thread>

#include "timer.hpp"
#include "syncedmem.hpp"
#include "gloo/transport/tcp/device.h"
#include "gloo/allreduce_ring.h"
#include "gloo/cuda_allreduce_ring_chunked.h"
#include "gloo/allreduce_ring_chunked.h"
#include "gloo/mpi/context.h"

using std::shared_ptr;
using std::vector;
using std::thread;
using std::bind;

void thread_func(vector<float*> data, int count, bool is_gpu, shared_ptr<gloo::mpi::Context>& context) {
	if (is_gpu) {
		for (size_t i = 0; i < data.size(); ++i) {
			gloo::CudaAllreduceRingChunked<float> allreduce(context, {data[i]}, count);
			allreduce.run();
		}
	}else{
		for (size_t i = 0; i < data.size(); ++i) {
			gloo::AllreduceRingChunked<float> allreduce(context, {data[i]}, count, gloo::ReductionFunction<float>::sum);
			allreduce.run();
		}
	}	
}

void multi_thread_performance(bool is_gpu) {
  auto dev = gloo::transport::tcp::CreateDevice("localhost");
  
  int thread_num = 4;    
  vector<shared_ptr<gloo::mpi::Context> > contexts(thread_num+1);
  for (int i = 0; i < thread_num+1; ++i) {
	contexts[i] = std::make_shared<gloo::mpi::Context>(MPI_COMM_WORLD);
	contexts[i]->connectFullMesh(dev);  
  }
  
  const int count = 1024*1024;
  const int P = 20;
  const int N = thread_num * P;
  vector<float*> data_vec(N);
  // SyncedMemory is a cpu&gpu memory manager which is copied from Caffe.
  SyncedMemory mem(N * count * sizeof(float));
  for (int i = 0; i < N; ++i) {
	if (is_gpu) {
		data_vec[i] = (float*)mem.mutable_gpu_data() + i * count;
	}else{
		data_vec[i] = (float*)mem.mutable_cpu_data() + i * count;
	}	
  }
  
  // I found the first time is running much slower, so run each context in advance.
  for (int i = 0; i < thread_num+1; ++i) {
	thread_func(data_vec, count, is_gpu, contexts[thread_num]);  
  }
  
  CPUTimer timer;
  timer.Start();
  thread_func(data_vec, count, is_gpu, contexts[thread_num]);
  float s_elapse = timer.MilliSeconds();
  
  vector<std::thread> threads;
  timer.Start();
  for (int i = 0; i < thread_num; ++i) {
	  vector<float*> data(data_vec.begin() + i*P, data_vec.begin() + (i+1)*P);
	  threads.push_back(std::thread(std::bind(thread_func, data, count, is_gpu, contexts[i])));
  }
  for (int i = 0; i < thread_num; ++i) {
	  threads[i].join();
  }
  float m_elapse = timer.MilliSeconds();

  int rank = contexts[0]->rank;
  if (0 == rank) {
	printf("is_gpu=%s, time for [%d] thread: %.2f ms, time for single thread: %.2f ms.\n", 
			is_gpu ? "true" : "false", thread_num, m_elapse, s_elapse);
  }
	  
}

int main(int argc, char** argv) {
  MPI_Init(&argc, &argv);
  
  multi_thread_performance(true);
  multi_thread_performance(false);
  
  MPI_Finalize();
  return 0;
}
```

**Comment by pietern** [comment]
A couple of observations: with 4 processes x 4 threads, are you 100% sure you're not overcommitting the CPU resources in your machine? With the size you're using (20 million floats) you end up spending quite some time reducing (summing) intermediate buffers on the CPU side. This is not cheap. You'll see much more benefit when you run many smaller (couple of kilobytes) allreduces in parallel. When you look at typical model size for vision models, it's on the order of 10's of millions. So running 4x20M on a single machine won't be representative for training your run of the mill vision model.

**Comment by pietern** [comment]
Closing this issue. If you have any follow up please feel free to post here on in a new issue.
